django/db/models/expressions.py

```diff
@@ -1109,12 +1130,14 @@ def as_oracle(self, compiler, connection, template=None, **extra_context):
         # CASE WHEN expression. Change the template since the When expression
         # requires a left hand side (column) to compare against.
         sql, params = self.as_sql(compiler, connection, template, **extra_context)
-        sql = 'CASE WHEN {} THEN 1 ELSE 0 END'.format(sql)
+        if not self.for_where:
+            sql = 'CASE WHEN {} THEN 1 ELSE 0 END'.format(sql)
```

Comment:
I'm curious. Is this the sole reason why `for_where` has to be passed around?

Reply:
I think so :neutral_face:

Reply:
Not trying to delay this even further but it seems a bit odd that 95% of the diff generated for this patch is to account for a weird handling of `EXISTS` on Oracle. I now realize that the `for_where` name is a bit misleading as it actually means _for use without a rhs in where_ because all expression annotations refered in a `filter` call will also be used in the `WHERE` clause.

Wouldn't `(CASE WHEN {} THEN 1 ELSE 0 END = 1)` work in both cases, or maybe that's just turning it back into a boolean expression? It's definitely not pretty and I haven't tested it but that would avoid a documented method signature change for dealing with a weird quirk of Oracle.

Another solution might be to add a database feature of the form `supports_select_boolean_expression` and have the compiler wrap unsupported backend selected expressions in `CASE WHEN {} THEN 1 ELSE 0 END`.

I'd be happy to work on a PR that does that this week-end to remove this Oracle hack and simplify this PR. Thoughts?

Reply:
Here's what I came up with https://github.com/django/django/pull/11641.

Reply:
It's looking much better! @charettes 
Do you have a better name in mind? I can't manage to come up with something better and simple.

Reply:
And a probably better alternative #11642

Reply:
@NyanKiyoshi if either of the above PR get merge we can completely remove the `for_where` kwarg juggling as it would become unnecessary.

Reply:
The whole addition of `for_where` and how ugly that seemed is IIRC the reason this stalled. Having a better approach makes me feel much happier!

Reply:
https://github.com/django/django/pull/11652 seems solid and should allow us to get rid of `for_where` 🎉 here.

Once `Expression.conditional` is defined it should be possible to get rid of the `XXX` regarding the currently disallowed `When(Exists)`.